Problem with int on display

### The problem

Hello

I have problem with error after esphome upgrade

=========================================

i spent many hours but i can’t find why i have such error - it works one year wihtout problem after upgrade of ESP i can’t recompile it

could you please help me :)
greetings
Adam

### Which version of ESPHome has the issue?

 2021.11.4

### What type of installation are you using?

Home Assistant Add-on

### Which version of Home Assistant has the issue?

2021.11.5

### What platform are you using?

ESP8266

### Board

_No response_

### Component causing the issue

display

### Example YAML snippet

```yaml
esphome:
  name: ogrod
  platform: ESP8266
  board: nodemcuv2

wifi:
  ssid: "music"
  password: "password"

  # Enable fallback hotspot (captive portal) in case wifi connection fails
  ap:
    ssid: "Ogrod Fallback Hotspot"
    password: "test"

captive_portal:

# Enable logging
logger:
# Enable Home Assistant API
api:
  password: "test"

ota:
  safe_mode: True
  password: test

font:
  - file: "pixelmix.ttf"
    id: digit_font
    size: 8

spi:
    mosi_pin: D7
    clk_pin: D5

display:
  - platform: max7219digit
    cs_pin: D8  
    num_chips: 4
    intensity: 5 
    lambda: |-
      it.printf(0, 0, id(digit_font), " %.1f°C  wilg. %.0f%%  %.1f hPa", id(bme280_temperature).state, id(bme280_humidity).state, id(bme280_pressure).state);
      it.strftime(200, 0, id(digit_font), " %H:%M ", id(hass_time).now());
      it.scroll(true, 1, 30, 10000, 400);

time:
  - platform: homeassistant
    timezone: Europe/Warsaw
    id: hass_time
    
    
i2c:
  sda: D2
  scl: D1
  scan: True
  id: bus_a
    
    
sensor:
  - platform: bme280
    temperature:
      name: "Ogrod BME280 Temperatura"
      id: bme280_temperature
    pressure:
      name: "Ogrod BME280 cisnienie"
      id: bme280_pressure
    humidity:
      name: "Ogrod BME280 Wilgotnosc"
      id: bme280_humidity
    address: 0x76
    update_interval: 120s

  - platform: template
    name: "Ogrod wysokosc"
    lambda: |-
      const float STANDARD_SEA_LEVEL_PRESSURE = 1013.25; //in hPa, see note
      return ((id(bme280_temperature).state + 273.15) / 0.0065) *
        (powf((STANDARD_SEA_LEVEL_PRESSURE / id(bme280_pressure).state), 0.190234) - 1); // in meter
    update_interval: 2015s
    icon: 'mdi:signal'
    unit_of_measurement: 'm'
    
  - platform: template
    name: "Ogrod Wilgotnosc Absolutna"
    lambda: |-
      const float mw = 18.01534;    // molar mass of water g/mol
      const float r = 8.31447215;   // Universal gas constant J/mol/K
      return (6.112 * powf(2.718281828, (17.67 * id(bme280_temperature).state)/(id(bme280_temperature).state + 243.5)) * id(bme280_humidity).state * mw)/((273.15 + id(bme280_temperature).state) * r); // in grams/m^3
    accuracy_decimals: 2
    update_interval: 360s
    icon: 'mdi:water'
    unit_of_measurement: 'g/m³'
  - platform: template
    name: "Ogrod Punkt Rosy"
    lambda: return (243.5*(log(id(bme280_humidity).state/100)+((17.67*id(bme280_temperature).state)/(243.5+id(bme280_temperature).state)))/(17.67-log(id(bme280_humidity).state/100)-((17.67*id(bme280_temperature).state)/(243.5+id(bme280_temperature).state))));
    unit_of_measurement: °C
    icon: 'mdi:thermometer-alert'
```


### Anything in the logs that might be useful for us?

```txt
/config/esphome/Ogrod.yaml: In lambda function:
/config/esphome/Ogrod.yaml:44:40: error: invalid conversion from 'int' to 'esphome::max7219digit::ScrollMode' [-fpermissive]
       it.scroll(true, 1, 30, 10000, 400);
                                        ^
In file included from src/esphome.h:27:0,
                 from src/main.cpp:3:
src/esphome/components/max7219digit/max7219digit.h:73:8: error:   initializing argument 2 of 'void esphome::max7219digit::MAX7219Component::scroll(bool, esphome::max7219digit::ScrollMode, uint16_t, uint16_t, uint16_t)' [-fpermissive]
   void scroll(bool on_off, ScrollMode mode, uint16_t speed, uint16_t delay, uint16_t dwell);
        ^
*** [/data/ogrod/.pioenvs/ogrod/src/main.cpp.o] Error 1
```


### Additional information

_No response_

I have come across what looks like the exact the same problem; (versions match those of the OP)
Here's the extract from the compile time log:

```
[snip]
Compiling /data/scrollerdisplay02/.pioenvs/scrollerdisplay02/lib975/ESPAsyncTCP-esphome/SyncClient.cpp.o
/config/esphome/scrollerdisplay02.yaml: In lambda function:
/config/esphome/scrollerdisplay02.yaml:36:38: error: invalid conversion from 'int' to 'esphome::max7219digit::ScrollMode' [-fpermissive]
       it.scroll(true, 1, 25, 50, 6000);
                                      ^
In file included from src/esphome.h:24:0,
                 from src/main.cpp:3:
src/esphome/components/max7219digit/max7219digit.h:73:8: error:   initializing argument 2 of 'void esphome::max7219digit::MAX7219Component::scroll(bool, esphome::max7219digit::ScrollMode, uint16_t, uint16_t, uint16_t)' [-fpermissive]
   void scroll(bool on_off, ScrollMode mode, uint16_t speed, uint16_t delay, uint16_t dwell);
        ^
Compiling /data/scrollerdisplay02/.pioenvs/scrollerdisplay02/lib975/ESPAsyncTCP-esphome/tcp_axtls.c.o
[snip]
```

And here's the corresponding section from the source of the device:

```
[snip]    
    intensity: 0
    lambda: |-
      it.scroll(true, 1, 25, 50, 6000);
      it.printf(0, 0, id(digit_font), "%.3f  %.3f  %.3f ms ", id(tripmin).state, id(tripavg).state, id(tripmax).state);
[snip]
sensor:
  - platform: uptime
    name: Scroller Display 2 Uptime
  - platform: wifi_signal
    name: Scroller Display 2 WiFi Signal
    update_interval: 60s
  - platform: homeassistant
    entity_id: sensor.ping_avg
    name: inetroundtripaverage
    id: tripavg
  - platform: homeassistant
    entity_id: sensor.ping_max
    name: inetroundtripmaximum
    id: tripmax
  - platform: homeassistant
    entity_id: sensor.ping_min
    name: inetroundtripminimum
    id: tripmin
[snip]

```

As the OP says, this has worked fine over the past year or so.  Just today I thought time to bring everything up to date.  Here's the log output from the device itself (obviously running the existing compiled code rather than a new version!): 

```
[snip]
INFO Reading configuration /config/esphome/scrollerdisplay02.yaml...
INFO Detected timezone 'Europe/London'
INFO Starting log output from scrollerdisplay02.local using esphome API
INFO Successfully connected to scrollerdisplay02.local
[16:18:16][I][app:102]: ESPHome version 2021.9.2 compiled on Oct  6 2021, 16:42:36
[16:18:17][C][wifi:501]: WiFi:
[16:18:17][C][wifi:361]:   SSID: [redacted]
[16:18:17][C][wifi:362]:   IP Address: 192.168.2.3
[16:18:17][C][wifi:364]:   BSSID: [redacted]
[16:18:17][C][wifi:365]:   Hostname: 'scrollerdisplay02'
[16:18:17][C][wifi:369]:   Signal strength: -50 dB ▂▄▆█
[16:18:17][C][wifi:373]:   Channel: 6
[16:18:17][C][wifi:374]:   Subnet: 255.255.252.0
[16:18:17][C][wifi:375]:   Gateway: 192.168.1.252
[16:18:17][C][wifi:376]:   DNS1: 192.168.1.252
[16:18:17][C][wifi:377]:   DNS2: (IP unset)
[16:18:17][C][spi:081]: SPI bus:
[16:18:17][C][spi:082]:   CLK Pin: GPIO16 (Mode: OUTPUT)
[16:18:17][C][spi:084]:   MOSI Pin: GPIO5 (Mode: OUTPUT)
[16:18:17][C][spi:085]:   Using HW SPI: NO
[16:18:17][C][uptime.sensor:030]: Uptime Sensor 'Scroller Display 2 Uptime'
[16:18:17][C][uptime.sensor:030]:   State Class: 'total_increasing'
[16:18:17][C][uptime.sensor:030]:   Unit of Measurement: 's'
[16:18:17][C][uptime.sensor:030]:   Accuracy Decimals: 0
[16:18:17][C][uptime.sensor:030]:   Icon: 'mdi:timer-outline'
[16:18:18][C][logger:193]: Logger:
[16:18:18][C][logger:194]:   Level: DEBUG
[16:18:18][C][logger:195]:   Log Baud Rate: 115200
[16:18:18][C][logger:196]:   Hardware UART: UART0
[16:18:18][C][homeassistant.time:010]: Home Assistant Time:
[16:18:18][C][homeassistant.time:011]:   Timezone: 'GMT0BST-1,M3.4.0/1,M10.5.0/2'
[16:18:18][C][version.text_sensor:021]: Version Text Sensor 'Scroller Display 2 ESPHome Version'
[16:18:18][C][version.text_sensor:021]:   Icon: 'mdi:new-box'
[16:18:18][C][max7219DIGIT:046]: MAX7219DIGIT:
[16:18:18][C][max7219DIGIT:047]:   Number of Chips: 16
[16:18:18][C][max7219DIGIT:048]:   Intensity: 0
[16:18:18][C][max7219DIGIT:049]:   Scroll Mode: 1
[16:18:18][C][max7219DIGIT:050]:   Scroll Speed: 25
[16:18:18][C][max7219DIGIT:051]:   Scroll Dwell: 6000
[16:18:18][C][max7219DIGIT:052]:   Scroll Delay: 50
[16:18:18][C][max7219DIGIT:053]:   CS Pin: GPIO15 (Mode: OUTPUT)
[16:18:18][C][max7219DIGIT:054]:   Update Interval: 0.5s
[16:18:18][C][captive_portal:148]: Captive Portal:
[16:18:18][C][web_server:152]: Web Server:
[16:18:18][C][web_server:153]:   Address: scrollerdisplay02.local:80
[16:18:18][D][homeassistant.sensor:024]: 'sensor.ping_avg': Got state 10.54
[16:18:18][D][sensor:121]: 'inetroundtripaverage': Sending state 10.53700  with 1 decimals of accuracy
[16:18:18][C][ota:029]: Over-The-Air Updates:
[16:18:18][C][ota:030]:   Address: scrollerdisplay02.local:8266
[16:18:18][D][homeassistant.sensor:024]: 'sensor.ping_max': Got state 11.62
[16:18:18][D][sensor:121]: 'inetroundtripmaximum': Sending state 11.62400  with 1 decimals of accuracy
[16:18:18][C][api:135]: API Server:
[16:18:18][C][api:136]:   Address: scrollerdisplay02.local:6053
[16:18:19][D][homeassistant.sensor:024]: 'sensor.ping_min': Got state 9.81
[16:18:19][D][sensor:121]: 'inetroundtripminimum': Sending state 9.80700  with 1 decimals of accuracy
[16:18:19][C][wifi_info:009]: WifiInfo IPAddress 'Scroller Display 2  IP'
[16:18:19][D][homeassistant.sensor:024]: 'sensor.ping_avg': Got state 10.54
[16:18:19][D][sensor:121]: 'inetroundtripaverage': Sending state 10.53700  with 1 decimals of accuracy
[16:18:19][C][wifi_info:010]: WifiInfo SSID 'Scroller Display 2 SSID'
[16:18:19][D][homeassistant.sensor:024]: 'sensor.ping_max': Got state 11.62
[16:18:19][D][sensor:121]: 'inetroundtripmaximum': Sending state 11.62400  with 1 decimals of accuracy
[16:18:19][C][wifi_info:011]: WifiInfo BSSID 'Scroller Display 2 BSSID'
[16:18:19][D][homeassistant.sensor:024]: 'sensor.ping_min': Got state 9.81
[16:18:19][D][sensor:121]: 'inetroundtripminimum': Sending state 9.80700  with 1 decimals of accuracy
[16:18:19][C][wifi_signal.sensor:009]: WiFi Signal 'Scroller Display 2 WiFi Signal'
[16:18:19][C][wifi_signal.sensor:009]:   Device Class: 'signal_strength'
[16:18:19][C][wifi_signal.sensor:009]:   State Class: 'measurement'
[16:18:19][C][wifi_signal.sensor:009]:   Unit of Measurement: 'dBm'
[16:18:19][C][wifi_signal.sensor:009]:   Accuracy Decimals: 0
[16:18:19][C][homeassistant.sensor:030]: Homeassistant Sensor 'inetroundtripaverage'
[16:18:19][C][homeassistant.sensor:030]:   State Class: ''
[16:18:19][C][homeassistant.sensor:030]:   Unit of Measurement: ''
[16:18:19][C][homeassistant.sensor:030]:   Accuracy Decimals: 1
[16:18:19][C][homeassistant.sensor:031]:   Entity ID: 'sensor.ping_avg'
[16:18:19][C][homeassistant.sensor:030]: Homeassistant Sensor 'inetroundtripmaximum'
[16:18:19][C][homeassistant.sensor:030]:   State Class: ''
[16:18:19][C][homeassistant.sensor:030]:   Unit of Measurement: ''
[16:18:19][C][homeassistant.sensor:030]:   Accuracy Decimals: 1
[16:18:19][C][homeassistant.sensor:031]:   Entity ID: 'sensor.ping_max'
[16:18:19][C][homeassistant.sensor:030]: Homeassistant Sensor 'inetroundtripminimum'
[16:18:19][C][homeassistant.sensor:030]:   State Class: ''
[16:18:19][C][homeassistant.sensor:030]:   Unit of Measurement: ''
[16:18:19][C][homeassistant.sensor:030]:   Accuracy Decimals: 1
[16:18:19][C][homeassistant.sensor:031]:   Entity ID: 'sensor.ping_min'
[16:18:39][D][homeassistant.sensor:024]: 'sensor.ping_avg': Got state 12.41
[16:18:39][D][sensor:121]: 'inetroundtripaverage': Sending state 12.40800  with 1 decimals of accuracy
[16:18:39][D][homeassistant.sensor:024]: 'sensor.ping_max': Got state 14.11
[16:18:39][D][sensor:121]: 'inetroundtripmaximum': Sending state 14.11400  with 1 decimals of accuracy
[16:18:39][D][homeassistant.sensor:024]: 'sensor.ping_min': Got state 9.75
[16:18:39][D][sensor:121]: 'inetroundtripminimum': Sending state 9.75300  with 1 decimals of accuracy
[snip]
```

--- a/components/display/max7219digit.rst
+++ b/components/display/max7219digit.rst
@@ -115,7 +115,7 @@ They can also be changed in the Lambda by adding the following command:
 
 
 - **on/off** -> switch scrolling on or off, use true or false
-- **mode** -> 0 = Continuous scrolling, 1 = Stop at end and reset
+- **mode** -> max7219digit::CONTINUOUS for continuous scrolling, max7219digit::STOP = Stop at end and reset
 - **speed** -> Set speed of scrolling (ms for every step of one dot)
 - **delay** -> pause time at start of scrolling
 - **dwell** -> pause at end of scrolling (only in mode 1)
@@ -127,9 +127,9 @@ They can also be changed in the Lambda by adding the following command:
         # ...
         lambda: |-
           # ...
-          it.scroll(true, 0, 100, 5000, 1500);
+          it.scroll(true, max7219digit::CONTINUOUS, 100, 5000, 1500);
           // OR
-          it.scroll(true, 0);
+          it.scroll(true, max7219digit::CONTINUOUS);
           // OR
           it.scroll(true);
 

